utils/errors/errors.go

```diff
@@ -0,0 +1,114 @@
+package errors
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/gogo/protobuf/grpc"
+	"github.com/tendermint/tendermint/libs/log"
+	"golang.org/x/xerrors"
+	grpc2 "google.golang.org/grpc"
+)
+
+type causer interface {
+	Cause() error
+}
+
+// Is checks if the error is a (wrapped) error of the given type
+func Is[T error](err error) bool {
+	switch e := err.(type) {
+	case nil:
+		return false
+	case T:
+		return true
+	case causer:
+		return Is[T](e.Cause())
+	case xerrors.Wrapper:
+		return Is[T](e.Unwrap())
+	default:
+		return false
+	}
+}
+
+// ServerWithSDKErrors wraps around a grpc server to return registered errors
+type ServerWithSDKErrors struct {
+	grpc.Server
+	Err    *sdkerrors.Error
+	Logger func(ctx sdk.Context) log.Logger
+}
+
+// RegisterService ensures that every server method that gets registered returns a registered error
+func (r ServerWithSDKErrors) RegisterService(sd *grpc2.ServiceDesc, server interface{}) {
+	for _, method := range sd.Methods {
+		method.Handler = func(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc2.UnaryServerInterceptor) (interface{}, error) {
+			res, err := method.Handler(srv, ctx, dec, interceptor)
+			if err != nil && !Is[*sdkerrors.Error](err) {
+				err = r.Err.Wrap(err.Error())
+				r.Logger(sdk.UnwrapSDKContext(ctx)).Debug(err.Error())
```

Comment:
Why not log sdk errors too?

Reply:
I didn't touch that code at all, it just got moved

Reply:
also this log is about knowing when an error had to be wrapped (hence the debug level), not about error logging. the calling function can handle that

Reply:
Hmm, I see